datafusion/scheduler/src/query.rs

```diff
@@ -0,0 +1,337 @@
+        self.output.is_closed()
+    }
+
+    /// Sends `output` to this query's output stream
+    pub fn send_query_output(&self, output: ArrowResult<RecordBatch>) {
+        let _ = self.output.unbounded_send(output);
+    }
+
+    /// Returns the [`Spawner`] associated with this [`Query`]
+    pub fn spawner(&self) -> &Spawner {
+        &self.spawner
+    }
+}
+
+/// When converting [`ExecutionPlan`] to [`Pipeline`] we may wish to group
+/// together multiple [`ExecutionPlan`], [`ExecGroup`] stores this state
+struct ExecGroup {
+    /// Where to route the output of the eventual [`Pipeline`]
+    output: Option<OutputLink>,
+
+    /// The [`ExecutionPlan`] from which to start recursing
+    root: Arc<dyn ExecutionPlan>,
+
+    /// The number of times to recurse into the [`ExecutionPlan`]'s children
+    depth: usize,
+}
+
+/// A utility struct to assist converting from [`ExecutionPlan`] to [`Query`]
+///
+/// The [`ExecutionPlan`] is visited in a depth-first fashion, gradually building
+/// up the [`RoutablePipeline`] for the [`Query`]. As nodes are visited depth-first,
+/// a node is visited only after its parent has been.
+struct QueryBuilder {
+    task_context: Arc<TaskContext>,
+    /// The current list of completed pipelines
+    in_progress: Vec<RoutablePipeline>,
+
+    /// A list of [`ExecutionPlan`] still to visit, along with
+    /// where they should route their output
+    to_visit: Vec<(Arc<dyn ExecutionPlan>, Option<OutputLink>)>,
+
+    /// Stores one or more [`ExecutionPlan`] to combine together into
+    /// a single [`ExecutionPipeline`]
+    exec_buffer: Option<ExecGroup>,
+}
+
+impl QueryBuilder {
+    fn new(plan: Arc<dyn ExecutionPlan>, task_context: Arc<TaskContext>) -> Self {
+        Self {
+            in_progress: vec![],
+            to_visit: vec![(plan, None)],
+            task_context,
+            exec_buffer: None,
+        }
+    }
+
+    /// Flush the current group of [`ExecutionPlan`] stored in `exec_buffer`
+    /// into a single [`ExecutionPipeline]
+    fn flush_exec(&mut self) -> Result<usize> {
+        let group = self.exec_buffer.take().unwrap();
+        let node_idx = self.in_progress.len();
+        self.in_progress.push(RoutablePipeline {
+            pipeline: Box::new(ExecutionPipeline::new(
+                group.root,
+                self.task_context.clone(),
+                group.depth,
+            )?),
+            output: group.output,
+        });
+        Ok(node_idx)
+    }
+
+    /// Visit a non-special cased [`ExecutionPlan`]
+    fn visit_exec(
+        &mut self,
+        plan: Arc<dyn ExecutionPlan>,
+        parent: Option<OutputLink>,
+    ) -> Result<()> {
+        let children = plan.children();
+
+        // Add the node to the current group of execution plan to be combined
+        // into a single [`ExecutionPipeline`].
+        //
+        // TODO: More sophisticated policy, just because we can combine them doesn't mean we should
```

Comment:
Can you explain what the tradeoffs are? At first it would seem to me that the larger a pipeline the better (as there is more work that can be done and so the scheduler will have the greatest chance of putting all the threads to use?

Reply:
I don't have a good grasp on them yet, this just replicates the grouping of the current parallelism story. This is a note to investigate if this is actually a good thing 😃

Reply:
I've reworked the terminology to refer to operators, rather than plans. I think this should make it more clear?